### PR TITLE
[Qt] reduce cs_main locks during tip update, more fluently update UI

### PR DESCRIPTION
**Pull request by jonasschnelli**
At the moment, the UI gets often not updated during initial sync or a catch-up of serval blocks. The UI's try_lock often can't acquire the lock which result in a bad user experience ("Baby, it's not updating!").

This PR changes the blocktip-update signal parameter form a bare hash (`uint256`) to a `CBlockIndex*`, which allows to safely get it's height and timestamp without locking anything. In addition, the blocktip-update signal now gets fired also during initial sync, but with an extra boolean flag that indicates the initial sync state.
The PR temporarily has a benchmark for the tip update signal. On my standard system it uses 0.03ms for the signal (=reasonable). Keep in mind that this signal is synchronous, but it will emit another asynchronous QT signal to the UI thread that will care about the UI update!

The timer based update (every 250ms) is still active for bandwidth and mempool UI observation, although there no cs_main lock is required.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Status bar during IBD looks like: https://bitcoin.jonasschnelli.ch/qt/statusbar.mov
Console: https://bitcoin.jonasschnelli.ch/qt/console.mov
(sorry the german lang.)

**Comment by laanwj** [comment]
Concept ACK, I've been wanting to do this for a long time

**Comment by jonasschnelli** [comment]
Time-profiled this PR on a mac and it looks like that the performance is very similar to current master.
I try to now to implement a time delta throttling instead of emitting the signal if `nHeight % 10 == 0` (only during initial sync).

Master:
bildschirmfoto 2015-11-27 um 11 06 55


PR:
bildschirmfoto 2015-11-27 um 11 05 40

**Comment by jonasschnelli** [comment]
Updated. Much smaller diff/changeset now.
Changes to a min time delta update to not over-emit signal to the UI during IBD on a fast computer. Now the UI gets updated if the last update was more then 250ms ago.

**Comment by laanwj** [comment]
This fixes #5664 for me. Even during catching-up and heavy verification, it keeps updating the block number consistently.

File: src/qt/clientmodel.cpp [hunk 1]
```diff
@@ -107,32 +104,8 @@ double ClientModel::getVerificationProgress() const
...
-        cachedBlockDate = newBlockDate;
-        cachedReindexing = fReindex;
-        cachedImporting = fImporting;
-
-        Q_EMIT numBlocksChanged(newNumBlocks, newBlockDate);
-    }
-
+    // no locking required at this point
+    // the following calls will aquire the required lock
     Q_EMIT mempoolSizeChanged(getMempoolSize(), getMempoolDynamicUsage());
```
Comment (laanwj) [inline src/qt/clientmodel.cpp]: Hm, spoke too soon. It does hang on long contention of cs_main, I think here.
The TRY_LOCK should be re-added here.

Comment (laanwj) [inline src/qt/clientmodel.cpp]: Hm no that can't be it, none of these require cs_main. Weird.

**Comment by laanwj** [comment]
Looks like it can still hang here:
```
#5  lock (this=<synthetic pointer>) at /store/orion/projects/bitcoin/experiment/boost/include/boost/thread/lock_types.hpp:346
#6  Enter (pszName=<optimized out>, pszFile=<optimized out>, nLine=<optimized out>, this=<synthetic pointer>) at ./sync.h:116
#7  CMutexLock (fTry=false, nLine=101, pszFile=0x5555559b7648 "qt/clientmodel.cpp", pszName=<optimized out>, mutexIn=..., this=<synthetic pointer>) at ./sync.h:137
#8  ClientModel::getVerificationProgress (this=<optimized out>) at qt/clientmodel.cpp:101
#9  0x00005555555db479 in BitcoinGUI::setNumBlocks (this=0x5555564ba3c0, count=375314, blockDate=...) at qt/bitcoingui.cpp:752
#10 0x000055555564960f in BitcoinGUI::qt_static_metacall (_o=0x5555564ba3c0, _c=<optimized out>, _id=<optimized out>, _a=<optimized out>) at qt/moc_bitcoingui.cpp:189
```
getVerificationProgress takes a lock on cs_main

**Comment by MarcoFalke** [comment]
Concept ACK. f60a931d9137854f3bbcbc0589ab31563662c3c3 Looks a lot smoother on my system.

**Comment by jonasschnelli** [comment]
Added another commit that solves the `getVerificationProgress()` locking issue.
The only reason why `getVerificationProgress()` needs a `cs_main` lock is because it accesses `chainActive.Tip()`.
The last commit will pass the signals `CBlockIndex*` to the `GuessVerificationProgress()` which make the whole UI update `cs_main` lock free.

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -672,7 +672,7 @@ void BitcoinGUI::setNumConnections(int count)
     labelConnectionsIcon->setToolTip(tr("%n active connection(s) to Bitcoin network", "", count));
 }
 
-void BitcoinGUI::setNumBlocks(int count, const QDateTime& blockDate)
+void BitcoinGUI::setNumBlocks(int count, const QDateTime& blockDate, const CBlockIndex *tip)
```
Comment (laanwj) [inline src/qt/bitcoingui.cpp]: I don't like passing the tip through the view code. Let's just pass the verification progress directly

Comment (jonasschnelli) [inline src/qt/bitcoingui.cpp]: Hmm... but i don't want to execute `getVerificationProgress()` during the synchronous core signal. I guess it's better to call this function over the UI thread.
But let me have a closer look...

Comment (laanwj) [inline src/qt/bitcoingui.cpp]: Hmm okay you have a point.
But ideally these kind of things happen in the model (ClientModel) in this case, not in the view class.
Don't like how core's handles leak all over the place now.
But yes maybe better to live with this for now...

**Comment by sipa** [comment]
@jonasschnelli @laanwj I benchmark GuessVerificationProgress (+random fetch from chainActive) here at around 80ns. I think it's perfectly acceptable to compute it before passing to the GUI.

**Comment by laanwj** [comment]
@sipa thanks; absolutely, in that case we should call it in the signal handler directly before sendingto the GUI. I expect `GetTimeMillis()` takes longer than that.

**Comment by jonasschnelli** [comment]
@sipa: Thanks for the info.

Just added another commit that calculate the verification progress during the synchronous core signal, and passes a double through the UI signal.
Feels very fast now (testes IBD, short catch-up 10 blocks).

File: src/qt/clientmodel.cpp [hunk 3]
```diff
@@ -99,40 +98,21 @@ size_t ClientModel::getMempoolDynamicUsage() const
...
 
-double ClientModel::getVerificationProgress() const
+double ClientModel::getVerificationProgress(const CBlockIndex *tipIn) const
 {
-    LOCK(cs_main);
-    return Checkpoints::GuessVerificationProgress(Params().Checkpoints(), chainActive.Tip());
+    CBlockIndex *tip = (CBlockIndex *)tipIn;
+    if (!tip)
+    {
+        LOCK(cs_main);
```
Comment (jonasschnelli) [inline src/qt/clientmodel.cpp]: This lock is only required when initially update the UI (first time, outside of the core signal scope).

**Comment by jonasschnelli** [comment]
Binares are built if someone likes to test this: https://bitcoin.jonasschnelli.ch/pulls/7112/

File: src/qt/clientmodel.cpp [hunk 4]
```diff
@@ -99,40 +98,21 @@ size_t ClientModel::getMempoolDynamicUsage() const
     return mempool.DynamicMemoryUsage();
 }
 
-double ClientModel::getVerificationProgress() const
+double ClientModel::getVerificationProgress(const CBlockIndex *tipIn) const
 {
-    LOCK(cs_main);
-    return Checkpoints::GuessVerificationProgress(Params().Checkpoints(), chainActive.Tip());
+    CBlockIndex *tip = (CBlockIndex *)tipIn;
```
Comment (sipa) [inline src/qt/clientmodel.cpp]: Don't recast to non-const, and certainly not with a C-style cast.

Comment (jonasschnelli) [inline src/qt/clientmodel.cpp]: Better like this (it's now `const_cast<CBlockIndex *>(tipIn)`)?
Or would it make more sense to pass a non-const through the core signal?

Comment (sipa) [inline src/qt/clientmodel.cpp]: No, I think GuessVerificationProgress should take a const CBlockIndex* :)

Comment (jonasschnelli) [inline src/qt/clientmodel.cpp]: Yeah. Would be nice.. but i guess this would be a too-broad change for this PR scope.

Comment (sipa) [inline src/qt/clientmodel.cpp]: Just change the argument to const there. No other changes needed (I tried) :)

File: src/main.cpp [hunk 5]
```diff
@@ -2402,8 +2402,12 @@ bool ActivateBestChain(CValidationState &state, const CBlock *pblock) {
             }
             // Notify external listeners about the new tip.
             GetMainSignals().UpdatedBlockTip(pindexNewTip);
-            uiInterface.NotifyBlockTip(hashNewTip);
         }
+        //TODO: remove tip notification benchmark
```
Comment (sipa) [inline src/main.cpp]: Maybe remove this again :)

File: src/main.cpp [hunk 6]
```diff
@@ -2636,9 +2636,11 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
             // Notify external listeners about the new tip.
             if (!vHashes.empty()) {
                 GetMainSignals().UpdatedBlockTip(pindexNewTip);
-                uiInterface.NotifyBlockTip(vHashes.front());
             }
         }
+        if (!vHashes.empty()) {
+            uiInterface.NotifyBlockTip(fInitialDownload, pindexNewTip);
+        }
```
Comment (jonasschnelli) [inline src/main.cpp]: @sdaftuar: I'm not familiar with the block announcements with headers, could you check if this rebase makes sense?

File: src/main.cpp [hunk 7]
```diff
@@ -2636,9 +2636,10 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
             // Notify external listeners about the new tip.
             if (!vHashes.empty()) {
                 GetMainSignals().UpdatedBlockTip(pindexNewTip);
-                uiInterface.NotifyBlockTip(vHashes.front());
             }
         }
+        // Always notify the UI if a new block tip was connected
+        uiInterface.NotifyBlockTip(fInitialDownload, pindexNewTip);
```
Comment (jonasschnelli) [inline src/main.cpp]: @sdaftuar: I'm not familiar with the block announcements with headers, could you check if this rebase makes sense?
I'm firing on this event even when `vHashes` is empty now.

**Comment by jonasschnelli** [comment]
Rebased. Removed notification benchmark.

**Comment by laanwj** [comment]
ACK. Works great now. The spinner is actually spinning again while syncing instead of blundering and glitching along :)

File: src/main.cpp (hunk 7 above)
Comment (sipa) [inline src/main.cpp]: No, you should only fire if vHashes is not empty. vHashes contains the
block hashes of blocks that were not in the best chain, but now are.

Comment (sipa) [inline src/main.cpp]: The code line is correct though: if vHashes is not empty, its first entry
will be the hash of the new tip anyway, so you can use pindexTip.

**Comment by jonasschnelli** [comment]
Added a commit that resolves conflicts with block header announcements (#7129) and now also includes PR #7037 (because it conflicts with it).

**Comment by sipa** [comment]
utACK non-GUI code.

**Pull request merged**